packages/native/types/base.d.ts

```diff
@@ -0,0 +1,227 @@
+}
+
+/**
+ * @typeparam ComponentProps  Props which will be included when withComponent is called
+ * @typeparam SpecificComponentProps  Props which will *not* be included when withComponent is called
+ */
+export type StyledComponent<
+  ComponentProps extends {},
+  SpecificComponentProps extends {} = {}
+> = NamedExoticComponent<ComponentProps & SpecificComponentProps> &
+  StyledWithComponent<ComponentProps>
+
+/**
+ * @typeparam ComponentProps  Props which will be included when withComponent is called
+ * @typeparam SpecificComponentProps  Props which will *not* be included when withComponent is called
+ * @typeparam StyleProps  Params passed to styles but not exposed as React props. These are normally library provided props
+ */
+export interface CreateStyledComponent<
+  ComponentProps extends {},
+  SpecificComponentProps extends {} = {},
+  StyleProps extends {} = {},
+  StyleType = ReactNativeStyle
+> {
+  /**
+   * @typeparam AdditionalProps  Additional props to add to your styled component
+   */
+  <AdditionalProps extends {} = {}>(
+    ...styles: ArrayInterpolation<
+      ComponentProps & SpecificComponentProps & StyleProps & AdditionalProps,
+      StyleType
+    >
+  ): StyledComponent<ComponentProps & AdditionalProps, SpecificComponentProps>
+  /**
+   * @typeparam AdditionalProps  Additional props to add to your styled component
+   */
+  <AdditionalProps extends {} = {}>(
+    template: TemplateStringsArray,
+    ...styles: ArrayInterpolation<
+      ComponentProps & SpecificComponentProps & StyleProps & AdditionalProps,
+      StyleType
+    >
+  ): StyledComponent<ComponentProps & AdditionalProps, SpecificComponentProps>
+}
+
+/**
+ * @desc
+ * This function accepts a React component.
+ *
+ * @example styled(MyComponent)({ width: 100 })
+ * @example styled(MyComponent)(myComponentProps => ({ width: myComponentProps.width })
+ * @example styled(View)({ width: 100 })
+ * @example styled(View)<Props>(props => ({ width: props.width })
+ */
+export interface CreateStyled<Theme extends {} = any> {
+  <
+    Component extends ComponentType<ComponentPropsWithoutRef<Component>>,
+    ForwardedProps extends keyof ComponentPropsWithoutRef<
+      Component
+    > = keyof ComponentPropsWithoutRef<Component>
+  >(
+    component: Component,
+    options: FilteringStyledOptions<
+      ComponentPropsWithoutRef<Component>,
+      ForwardedProps
+    >
+  ): CreateStyledComponent<
+    Pick<ComponentPropsWithoutRef<Component>, ForwardedProps> & {
+      theme?: Theme
+    },
+    {},
+    { theme: Theme },
+    ReactNativeStyleType<ComponentPropsWithoutRef<Component>>
+  >
+
+  <Component extends ComponentType<ComponentPropsWithoutRef<Component>>>(
+    component: Component,
+    options?: StyledOptions<ComponentPropsWithoutRef<Component>>
+  ): CreateStyledComponent<
+    ComponentPropsWithoutRef<Component> & { theme?: Theme },
+    {},
+    { theme: Theme },
+    ReactNativeStyleType<ComponentPropsWithoutRef<Component>>
```

Comment:
as long as the custom component defines a `style?: StyleProp<T>`, the `T` will get inferred and applied to the template interpolation types.